File: snakesolver.c

// Compile with:
// gcc -std=c99 snakesolver.c && ./a.out

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#define DIM 4
int snake_short[]= { 5, 1, 1, 2, 1, 1, 1, 6, 1, 1, 2, 2, 4, 2, 3, 4, 3, 5, 5, 2, 2, 4, 2, 2, 2, 0 };
// int snake_short="7 1 1 2 1 1 1 6 1 1 2 2 4 2 3 4 3 5 5 2 2 4 2 2";
// int snake_short="2 4 2 2 4 4 4 2 2 4 1 1 4 2 2 4 4 4 2 2 4 3 1";
const int exit_after_first_solution= 1;

#define DIMM1 (DIM - 1)
#define DIMP1 (DIM + 1)
#define DIMP2 (DIM + 2)
#define DIMP2SQ (DIMP2 * DIMP2)
#define DIMP2CB (DIMP2 * DIMP2 * DIMP2)
#define SLEN (DIM * DIM * DIM)

#define xyz2n(x, y, z) (z * DIMP2SQ + y * DIMP2 + x )
#define n2x(n) (n % DIMP2)
#define n2y(n) ((n / DIMP2) % DIMP2)
#define n2z(n) ((n / DIMP2SQ) % DIMP2)

int result[DIMP2CB];
int solved[DIMP2CB];

int snake[SLEN];
int offset=0;
int tries= 0;

char *get_cube() {
    static char buf[10000];
    char *bufp= buf;

    for ( int z= 1; z <= DIM; z++ ) {
        bufp += sprintf(bufp, "z=%d\n"
            "\t+----+----+----+----+"
            "\t+---+---+---+---+\n", z);
        for ( int y= 1; y <= DIM; y++ ) {
            bufp += sprintf(bufp, "\t|");
            for ( int x= 1; x <= DIM; x++ ) {
                int n= xyz2n(x, y, z);
                bufp += sprintf(bufp, " %2d |", result[n]);
            }
            bufp += sprintf(bufp, "\t|");
            for ( int x= 1; x <= DIM; x++ ) {
                int n= xyz2n(x, y, z);
                int snake_i= result[n];
                if ( snake_i < 0 ) {
                    bufp += sprintf(bufp, " ? |");
                }
                else if ( snake[snake_i] == 1 ) {
                    bufp += sprintf(bufp, " X |");
                }
                else {
                    bufp += sprintf(bufp, " X |");
                }
            }
            bufp += sprintf(bufp, ""
                "\n\t+----+----+----+----+"
                "\t+---+---+---+---+\n");
        }
        bufp += sprintf(bufp, "\n");
    }

    return buf;
}

void print_cube() {
    puts(get_cube());
}


void has_solved() {
    printf("\n\nSOLVED with %d tries\n", tries);
    print_cube();
    if ( exit_after_first_solution ) {
        exit(1);
    }
}

void find_way( int n, int idx, int dir, int fin ) {

    if ( fin ) {
        if ( result[n] == 0 ) { // naechstes element ist der anfang der schlange
            has_solved();
        }
        return;
    }

//    printf("fin=%d idx=%d dir=%d n=%d result=%d solved=%d\n", fin, idx, dir, n, result[n], solved[n]);

    if ( result[n] > -2 ) return;
    if ( solved[n] > 0 && solved[n] != snake[idx] ) return; // color cares but does not match

//    printf("fin=%d idx=%d dir=%d n=%d\n", fin, idx, dir, n);

    tries++;
  
    if ( tries % 1000000 == 0 ) {
        printf("\rtries: %d (%d,%d,%d) -> idx: %d\n", tries, n2x(n), n2y(n), n2z(n), snake[idx]);
    }

    result[n]= idx++;

    fin= idx >= SLEN;

    if ( dir != 1 ) find_way(n + 1,       idx, 0, fin);
    if ( dir != 0 ) find_way(n - 1,       idx, 1, fin);
    if ( dir != 3 ) find_way(n + DIMP2,   idx, 2, fin);
    if ( dir != 2 ) find_way(n - DIMP2,   idx, 3, fin);
    if ( dir != 5 ) find_way(n + DIMP2SQ, idx, 4, fin);
    if ( dir != 4 ) find_way(n - DIMP2SQ, idx, 5, fin);

    result[n]= -2;
}

int main () {

    int color= 2;

    int j= 0;
    for ( int i= 0; snake_short[i]; i++ ) {
        color= 3 - color;
        while ( snake_short[i]-- ) snake[j++]= color;
    }

    assert( j == SLEN );
    assert( (j & 1) == 0 );

    for ( int x= 0; x <= DIMP1; x++ ) {
        for ( int y= 0; y <= DIMP1; y++ ) {
            for ( int z= 0; z <= DIMP1; z++ ) {
                int n= xyz2n(x, y, z);
                if ( x == 0 || y == 0 || z == 0
                    || x == DIMP1 || y == DIMP1 || z == DIMP1
                ) {
                    result[n]= -1;
                    solved[n]= 0;  // Unused
                }
                else if ( x == 1 || y == 1 || z == 1
                    || x == DIM || y == DIM || z == DIM
                ) {
                    result[n]= -2;
                    solved[n]= (((x - 1) >> 1) + ((y - 1) >> 1) + ((z - 1) >> 1)) % 2 + 1;
                }
                else {
                    result[n]= -2;
                    solved[n]= 0;
                }
            }
        }
    }

    while ( offset < SLEN ) {
        if ( offset ) {
            int last= snake[SLEN - 1];
            for ( int i= SLEN - 1; i > 0; i-- ) snake[i]= snake[i - 1];
            snake[0]= last;
            printf("\nstarting over (offset %d)\n", offset);
        }
        offset++;

        find_way(xyz2n(1, 1, 1), 0, -1, 0);
    }

    // print_cube();

    return 0;
}

File: snakesolver.scala

object Main1 {
    
    val dim = 4
    val snake_short = List(
        5, 1, 1, 2, 1, 1, 1, 6, 1, 1, 2, 2, 4, 2, 3, 4, 3, 5, 5, 2, 2, 4, 2, 2, 2
    )
    val exit_after_first_solution = true
        
    val dimm1 = dim - 1
    val dimp1 = dim + 1
    val dimp2 = dim + 2
    val dimp2sq = dimp2 * dimp2
    val dimp2cb = dimp2 * dimp2 * dimp2
    
    val slen= dim * dim * dim

    def xyz2n(x: Int, y: Int, z: Int) = z * dimp2sq + y * dimp2 + x
    def n2x(n: Int) = n % dimp2
    def n2y(n: Int) = (n / dimp2) % dimp2
    def n2z(n: Int) = (n / dimp2sq) % dimp2

    var snake = List[Int]()

    def expand( list: List[Int], start: Int ): List[Int]= {
        def expand1( value: Int, n: Int ): List[Int] = {
            if (n == 0) Nil
            else value :: expand1( value, n-1 )
        }

        if (list == Nil) Nil
        else expand1( start, list.head ) ::: expand( list.tail, 3 - start)
    }   

    var offset: Int = 0
    var tries: Int = 0
 
    var result= new Array[Int](dimp2cb)
    var solved= new Array[Int](dimp2cb)

    def init = {
        for (z <- List.range(0, dimp2)) {
            for (y <- List.range(0, dimp2)) {
                for (x <- List.range(0, dimp2)) {
                    val n= xyz2n(x, y, z)
                    // println( List(x, y, z, n) )

                    if (x == 0 || y == 0 || z == 0
                        || x == dimp1 || y == dimp1 || z == dimp1
                    ) {
                        result(n)= -2
                        solved(n)= 0
                    }
                    else if (x == 1 || y == 1 || z == 1
                        || x == dim || y == dim || z == dim
                    ) {
                        result(n)= -1
                        solved(n)= (((x - 1) >> 1) + ((y - 1) >> 1) + ((z - 1) >> 1)) % 2 + 1
                    }
                    else {
                        result(n)= -1
                        solved(n)= 0
                    }
                }
            }
        }
    }

    def get_cube = {
        var r = "";

        for (z <- List.range(1, dimp1)) {

            r += ("z=" + z + "\n")
            r += "\t+----+----+----+----+"
            r += "\t+---+---+---+---+\n"
         
            for (y <- List.range(1, dimp1)) {
             
                r += "\t|"

                for (x <- List.range(1, dimp1)) {
                    val n= xyz2n(x, y, z)
                    // r += String.format(" %2d |", Array(result(n)))
                    r += " " + result(n) + " |"
                }

                r += "\t|"

                for (x <- List.range(1, dimp1)) {
                    val n= xyz2n(x, y, z)
                    r= r + " " + (if (snake(result(n)) == 1) "X" else " ") + " |"
                }

                r += "\n"
            }
            r += "\t+----+----+----+----+"
            r += "\t+---+---+---+---+\n"
        }
        r += "\n"
        r
    }

    def print_cube = {
        print( get_cube )
    }

    def print_solved {
        print( "\n\nSOLVED with " + tries + " tries\n" )
        print_cube
        
        if (exit_after_first_solution) System.exit(1)
    }
    
    def find_way(n: Int, idx: Int, dir: Int, fin: Boolean): Unit = {

        if (fin) {
            if (result(n) == 0) print_solved
        }
        else if (result(n) != -1) {
        }
        else if (solved(n) != 0 && solved(n) != snake(idx)) {
        }
        else {
            tries += 1;

            if (tries % 300000 == 0)
                println( "tries: " + tries + " (" + n2x(n) + "," + n2y(n) + "," + n2z(n) + ") -> idx: " + snake(idx))

            result(n)= idx;
            val newIdx = idx + 1
            val newFin= newIdx >= snake.length

            if (dir != 1) find_way(n + 1,       newIdx, 0, newFin)
            if (dir != 0) find_way(n - 1,       newIdx, 1, newFin)
            if (dir != 3) find_way(n + dimp2,   newIdx, 2, newFin)
            if (dir != 2) find_way(n - dimp2,   newIdx, 3, newFin)
            if (dir != 5) find_way(n + dimp2sq, newIdx, 4, newFin)
            if (dir != 4) find_way(n - dimp2sq, newIdx, 5, newFin)

            result(n)= -1
        }
    }
    
    def solve = {
        while (offset < slen) {
            if (offset > 0) {
                 snake= snake.tail ::: List(snake.head)
                 print( "\nstarting over (" + offset + ")\n" )
            }
            offset += 1
            find_way(xyz2n(1, 1, 1), 0, -1, false)
        }
    }
    
    def main(args: Array[String]): Unit = {
        snake= expand(snake_short, 1)
        init
        solve
    }
}
